contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go

```diff
@@ -697,9 +681,88 @@ func deleteInstanceProfile(instanceProfileID *string, iamClient *iam.IAM, logger
+	describeRoleInput := iam.GetRoleInput{}
+	describeRoleInput.RoleName = &roleName
+	iamClient := iam.New(session)
+	if _, err := iamClient.GetRole(&describeRoleInput); err != nil && err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
+		return err
+	}
+
+	// empty the role
+	logger.Debugf("deleting policies from role: %s", roleName)
+	if err := deletePoliciesFromRole(&roleName, iamClient); err != nil && err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
+		logger.Debugf("error deleting policies from role: %v", err)
+		return err
+	}
+	describeProfileInput := iam.GetInstanceProfileInput{}
+	describeProfileInput.InstanceProfileName = &profileName
+	if _, err := iamClient.GetInstanceProfile(&describeProfileInput); err != nil && err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
+		return err
+	}
+
+	// detach role from profile
+	logger.Debugf("detaching role from profile: %s", profileName)
+	_, err := iamClient.RemoveRoleFromInstanceProfile(&iam.RemoveRoleFromInstanceProfileInput{
+		InstanceProfileName: &profileName,
+		RoleName:            &roleName,
+	})
+	if err != nil && err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
+		logger.Debugf("error removing role from instance profile: %v", err)
+		return err
+	}
+	if err == nil {
+		logger.Infof("Removed role %v from instance profile %v", roleName, profileName)
+	}
+	// delete profile
+	logger.Debugf("deleting instance profile: %v", profileName)
+	_, err = iamClient.DeleteInstanceProfile(&iam.DeleteInstanceProfileInput{
+		InstanceProfileName: &profileName,
+	})
+	if err != nil && err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
+		logger.Debugf("error deleting instance profile %s: %v", profileName, err)
+		return err
+	}
+	if err == nil {
+		logger.Debugf("deleted profile %s", profileName)
+	}
+	// now we can delete role
+	logger.Debugf("deleted policies from role %s", roleName)
+	deleteRoleInput := iam.DeleteRoleInput{}
+	deleteRoleInput.RoleName = &roleName
+	if _, err := iamClient.DeleteRole(&deleteRoleInput); err != nil && err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
+		logger.Debugf("error deleting role %s: %v", roleName, err)
+		return err
+	}
+	if err == nil {
+		logger.Debugf("deleted role %s", roleName)
+	}
+	return nil
+}
+
+// deleteIAMresources will delete any IAM resources created by the installer that are not associated with a running instance
+// Currently openshift/installer creates 3 roles per cluster, 1 for master|worker|bootstrap and identified by the
+// cluster name used to install the cluster.
+func deleteIAMresources(session *session.Session, filter awsFilter, clusterName string, logger log.FieldLogger) (bool, error) {
+	logger.Debugf("Deleting IAM resources")
+	defer logger.Debugf("Exiting deleting IAM resources")
+	installer_type := []string{"master", "worker", "bootstrap"}
+	for _, t := range installer_type {
+		roleName := fmt.Sprintf("%s-%s-role", clusterName, t)
```

Comment:
can we put a comment here explaining the expected naming scheme for the various IAM objects?